src/rust/pants_ng/options/src/pants_invocation_tests.rs

```diff
@@ -0,0 +1,367 @@
+            global_flags: vec![],
+            commands: vec![mk_command("foo2-bar", vec![], None,)],
+            specs: vec![],
+            passthru: None,
+        },
+    );
+}
+
+#[test]
+fn test_command_and_subcommand() {
+    assert_eq!(
+        mk_invocation("pants --foo --bar=baz cmd subcmd").unwrap(),
+        PantsInvocation {
+            global_flags: vec![mk_flag("foo", None), mk_flag("bar", Some("baz")),],
+            commands: vec![mk_command(
+                "cmd",
+                vec![],
+                Some(mk_subcommand("subcmd", vec![]))
+            )],
+            specs: vec![],
+            passthru: None,
+        },
+    );
+
+    assert_eq!(
+        mk_invocation("pants --foo cmd --bar=baz subcmd").unwrap(),
+        PantsInvocation {
+            global_flags: vec![mk_flag("foo", None),],
+            commands: vec![mk_command(
+                "cmd",
+                vec![mk_flag("bar", Some("baz")),],
+                Some(mk_subcommand("subcmd", vec![]))
+            )],
+            specs: vec![],
+            passthru: None,
+        },
+    );
+
+    assert_eq!(
+        mk_invocation("pants --foo cmd --bar=baz subcmd --qux=quux").unwrap(),
+        PantsInvocation {
+            global_flags: vec![mk_flag("foo", None),],
+            commands: vec![mk_command(
+                "cmd",
+                vec![mk_flag("bar", Some("baz")),],
+                Some(mk_subcommand("subcmd", vec![mk_flag("qux", Some("quux"))]))
+            )],
+            specs: vec![],
+            passthru: None,
+        },
+    );
+
+    assert_eq!(
+        mk_invocation("pants --foo cmd --bar=baz subcmd --qux=quux path/to/file1 ./path2").unwrap(),
+        PantsInvocation {
+            global_flags: vec![mk_flag("foo", None),],
+            commands: vec![mk_command(
+                "cmd",
+                vec![mk_flag("bar", Some("baz")),],
+                Some(mk_subcommand("subcmd", vec![mk_flag("qux", Some("quux"))]))
+            )],
+            specs: vec!["path/to/file1".to_string(), "./path2".to_string()],
+            passthru: None,
+        },
+    );
+
+    assert_eq!(
+        mk_invocation("pants --foo cmd --bar=baz subcmd --qux=quux path/to/file1 ./path2 -- passthru_arg --passthru-flag").unwrap(),
+        PantsInvocation {
+            global_flags: vec![mk_flag("foo", None),],
+            commands: vec![mk_command(
+                "cmd",
+                vec![mk_flag("bar", Some("baz")),],
+                Some(mk_subcommand("subcmd", vec![mk_flag("qux", Some("quux"))]))
+            )],
+            specs: vec!["path/to/file1".to_string(), "./path2".to_string()],
+            passthru: Some(vec![
+                "passthru_arg".to_string(),
+                "--passthru-flag".to_string()
+            ]),
+        },
+    );
+}
+
+#[test]
+fn test_multiple_commands_and_subcommands() {
+    assert_eq!(
+        mk_invocation("pants cmd1 subcmd1 + cmd2 + cmd3 subcmd3").unwrap(),
```

Comment:
So obviously `|` as a character is out, but I'm wondering if these should be thought of as "do all of this and then do the next one", or "maybe 'line' by 'line' depending on the command".

Thinking of #10542 for example

Reply:
I think it would be best if these were always concurrent, no? You can get sequential behavior by running `pants` multiple times.